common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -106,28 +127,32 @@ public MaterialConfig config() {
     }
 
     public List<Modification> modificationsSince(File baseDir, Revision revision, final SubprocessExecutionContext execCtx) {
+        unshallowIfNeeded(ProcessOutputStreamConsumer.inMemoryConsumer(), revision, baseDir);
         GitCommand gitCommand = getGit(baseDir);
         return gitCommand.modificationsSince(revision);
     }
 
     public MaterialInstance createMaterialInstance() {
-        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString());
+        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString(), shallowClone);
     }
 
     @Override
     protected void appendCriteria(Map<String, Object> parameters) {
         parameters.put(ScmMaterialConfig.URL, url.forCommandline());
         parameters.put("branch", branch);
+        parameters.put("shallowClone", shallowClone);
```

Comment:
hi @arvindsv:

> My point is, why can't this value be like submoduleFolder which is in GitMaterial? It's not part of the fingerprint, but is persisted.

If user have two pipeline point to same git repo and branch, and they turn on or off shallow clone in one of them, what is the behavior should we expect if we do not distinguish them?  When I was reading the code, I do wonder how 'submoduleFolder' handling avoids this confusing when they have inconsistent settings.

>Also, as a part of pluggable SCM support, a new field called "additionalData" was introduced. It can be used in any material and can hold a JSON value. Is that something we should look into for this?

I was looking into it. 'additionalData' currently has very specific usage for storing additional value coming along with Pluggable SCM plugin polling result. I am not sure I want to mess with it. Plus it does not help solving the material identity problem.

>So, even a GitMaterial row in it has a null value for a column called useTickets, which is a Perforce (@jyotisingh's favorite SCM) field

Yes, once in a STI hell, very hard to get out :-). But it is not that bad in GO's case because MaterialInstance model is almost logic free.

>  But, the Go Server is not the goal of this PR. The agents are the goal (where a shallow or non-shallow checkout really matters).

I totally agree. I feel in a ideal implementations, flyweight clone on go server should always be shallow (because historical revision is never used), and shallow flag is turn on and off for agents. But that would involve too much changes to the existing process.